### DISCUSSION TITLE
Global Secondary Index not working

### DISCUSSION THREAD

**Discussion by Wheredidigo** [sensedeep/dynamodb-onetable#155] [Q&A] [answered]
I'm trying to use a .find off of a Model using the index: gs1, but I'm not getting any results even though I know the data is in there.

Here is part of my schema where I define the model I'm trying to find:

```
export default {
  version: '0.0.1',
  indexes: {
    primary: { hash: 'pk', sort: 'sk' },
    gs1: { hash: 'gs1pk', sort: 'gs1sk', project: 'all' },
  },
  models: {
    Event: {
      pk: { type: String, value: '${_type}#${id}' },
      sk: { type: String, value: '${_type}#' },
      id: { type: String, uuid: true, required: true, validate: Match.uuid },
      eventName: { type: String, required: true },
      domainName: { type: String, enum: ['DataPlatform', 'Data', 'Order'], required: true },
      companyName: { type: String, enum: ['1800Contacts', 'Liingo'], required: true },
      isPrivate: { type: Boolean, required: true },
      eventVersion: { type: Number, required: true },

      // Search by domainName, and type
      gs1pk: { type: String, value: '${_type}#' },
      gs1sk: { type: String, value: '${domainName}#${id}' },

      // Search by companyName, and type
      // gs2pk: { type: String, value: '${_type}#' },
      // gs2sk: { type: String, value: '${companyName}#${id}' },
    },
```

And here is my method where I'm trying to get all of the "Event" models back by the gs1 index:

```
  async getEvents(): Promise<Array<EventModel>> {
    const events = new Array<EventModel>();
    type EventType = Entity<typeof Schema.models.Event>;
    const eventModel = this.dynamoDbTable.getModel<EventType>('Event');

    const eventStorages = await eventModel.find({ gs1pk: 'Event#' }, { index: 'gs1' });
    console.log(eventStorages);

    for (const eventStorage of eventStorages) {
      const newEventModel = new EventModel();

      newEventModel.id = eventStorage.id;
      newEventModel.eventName = eventStorage.eventName;
      newEventModel.domainName = DomainEnum[this.getEnumKeyByEnumValue(DomainEnum, eventStorage.domainName)];
      newEventModel.companyName = CompanyEnum[this.getEnumKeyByEnumValue(CompanyEnum, eventStorage.companyName)];
      newEventModel.version = eventStorage.eventVersion;
      newEventModel.isPrivate = eventStorage.isPrivate;

      events.push(newEventModel);
    }

    return events;
  }
```

Whenever I hit this method and check the log, I'm seeing it log an empty array.

When I go into the DynamoDb console and run a query against the gs1 index, I can see the 3 records I have in ther
![Screen Shot 2021-10-14 at 10 21 34 AM](https://user-images.githubusercontent.com/12992257/137357768-225ce791-d9f8-4888-bf77-e8d6dec5df66.png)
e.

**Comment by mobsense** [comment, marked as answer]
First, add {log: true} to the API params and setup a logger as described in the README. Then you can see the actual generated DynamoDB command.

Because your model says:

```
      gs1pk: { type: String, value: '${_type}#' },
      gs1sk: { type: String, value: '${domainName}#${id}' },
```

When you do a find, you don't typically provide the gs1pk and gs1sk properties to the API. Instead you provide the domainName and ID. The gs1pk and gs1sk values are determined from the type, domainName and id values you provide.

If you don't provide the domainName and ID, find() will keep the literal defined leading portion and use that in a beginsWith clause. In your case, you don't have a leading fixed portion in your gs1sk. AND you've hit a bug where the actual gs1sk used is (I think) beginsWith "#". This of course will fail.

We've committed a fix for 1.9.2 which will fix this.

In the meantime, as a workaround. set gs1sk to null in your API.

    await eventModel.find({ gs1sk: null }, { index: 'gs1' });

**Reply by Wheredidigo** [reply]
Sorry this took me so long to update, but what I ended up doing was this:

We updated the gs1pk and gs1sk to this:

```
     // Search by domainName, and type
      gs1pk: { type: String, value: '${_type}#' },
      gs1sk: { type: String, value: '${_type}#${domainName}#${id}' },
```

And then we figured out how to do the correct "BeginsWith" in typescript by utilizing tunnel:

```
    const events = new Array<EventModel>();
    const eventStorages = await this.eventModel.find({ gs1pk: 'Event#' }, { index: 'gs1', tunnel: { begins: { gs1sk: 'Event' } } });
```

You're right though that what we found after implementing the logging was that the way we were doing it before was making the query pass in a BeginsWith that wasn't correct.

**Reply by mobsense** [reply]
Great. We've got a patch coming that supports "begins" for TypeScript without needing tunnel. Should be in the next release. Thanks to @markandrus

**Comment by harinair** [comment]
I struggled a lot to figure out how to do find by index. This helped me a lot. Why is this not in documentation?

This is a very powerful product by the documentation is significantly lagging, especially expert features.

**Reply by mobsense** [reply]
Thanks for the feedback.  Do you mean how to use {index: 'name'} or how to do find using begins?

Happy to update the doc. If you can give a bit more detail about exactly what you'd like to see updated, we are happy to oblige.

The doc is pretty low level, what kind of expert features would you like to see documented better?
